internal/engine/planner/merge.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021-2025 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package planner
+
+import (
+	"maps"
+	"slices"
+	"sort"
+
+	enginev1 "github.com/cerbos/cerbos/api/genpb/cerbos/engine/v1"
+	"github.com/cerbos/cerbos/internal/conditions"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+func Merge(outputs map[string]*enginev1.PlanResourcesOutput) *enginev1.PlanResourcesFilter {
+	response := &enginev1.PlanResourcesFilter{
+		Kind: enginev1.PlanResourcesFilter_KIND_CONDITIONAL,
+	}
+	scopes := slices.Sorted(maps.Keys(outputs))
+	expressions := make([]*enginev1.PlanResourcesFilter_Expression_Operand, 0, len(outputs))
+	nots := make([]string, 0, len(outputs))
+	conds := make(map[string][]string, len(outputs))
+	for _, scope := range scopes {
+		output := outputs[scope]
+		switch output.Filter.Kind {
+		case enginev1.PlanResourcesFilter_KIND_ALWAYS_ALLOWED:
+			expressions = append(expressions, mkScopeOp(Equals, scope))
+		case enginev1.PlanResourcesFilter_KIND_ALWAYS_DENIED:
+			nots = append(nots, scope)
+		case enginev1.PlanResourcesFilter_KIND_CONDITIONAL:
+			conds[output.FilterDebug] = append(conds[output.FilterDebug], scope)
+		case enginev1.PlanResourcesFilter_KIND_UNSPECIFIED:
+			panic("unspecified filter kind")
```

Comment:
Let's return an error here instead of panicking because it's easy to forget to update this code path during a change and it'd go unnoticed for a while until someone hits it in production.